src/PostgREST/App.hs

```diff
@@ -252,47 +239,61 @@ app dbStructure conf apiRequest =
+    in (status, contentRange)
+
   mapSnd f (a, b) = (a, f b)
   readDbRequest = DbRead <$> buildReadRequest (configMaxRows conf) (dbRelations dbStructure) (map (mapSnd pdReturnType) $ dbProcs dbStructure) apiRequest
   mutateDbRequest = DbMutate <$> buildMutateRequest apiRequest
   selectQuery = requestToQuery schema False <$> readDbRequest
-  countQuery = requestToCountQuery schema <$> readDbRequest
-  mutateQuery = requestToQuery schema False <$> mutateDbRequest
-  readSqlParts = (,) <$> selectQuery <*> countQuery
-  mutateSqlParts = (,) <$> selectQuery <*> mutateQuery
-  respondToRange response = if topLevelRange == emptyRange
-                            then return $ errResponse status416 "HTTP Range error"
-                            else response
-  rangeHeader queryTotal tableTotal = let lower = rangeOffset topLevelRange
-                                          upper = lower + toInteger queryTotal - 1
-                                          contentRange = contentRangeH lower upper (toInteger <$> tableTotal)
-                                          status = rangeStatus lower upper (toInteger <$> tableTotal)
-                                      in (status, contentRange)
-
-
-serves :: Monad m => [ContentType] -> [ContentType] ->
-  (ContentType -> m Response) -> m Response
-serves sProduces cAccepts resp =
-  case mutuallyAgreeable sProduces cAccepts of
-    Nothing -> do
-      let failed = intercalate ", " $ map (toS . toHeader) cAccepts
-      return $ errResponse status415 $
-        "None of these Content-Types are available: " <> failed
-    Just ct -> resp ct
+  servesMutateRequest = servesDbRequest selectQuery $ requestToQuery schema False <$> mutateDbRequest
+  servesReadRequest resp =
+    if topLevelRange == emptyRange
+      then return $ errResponse status416 "HTTP Range error"
+      else servesDbRequest selectQuery (requestToCountQuery schema <$> readDbRequest) resp
+  servesContentTypesForRequest = servesMatchingContentTypes (iAccepts apiRequest) (iAction apiRequest)
+
+servesDbRequest :: Monad m => Either Text SqlQuery -> Either Text SqlQuery -> (SqlQuery -> SqlQuery -> m Response) -> m Response
+servesDbRequest selectQuery countQuery resp =
+  case (,) <$> selectQuery <*> countQuery of
+    Left e -> return $ responseLBS status400 [toHeader CTApplicationJSON] $ toS e
+    Right (sq,q) -> resp sq q
+
+servesMatchingContentTypes :: Monad m => [ContentType] -> Action -> (ContentType -> m Response) -> m Response
+servesMatchingContentTypes accepts action = serves contentTypesForRequest accepts
+  where
+    contentTypesForRequest =
+      case action of
+        ActionRead   -> [CTApplicationJSON, CTTextCSV]
+        ActionCreate -> [CTApplicationJSON, CTTextCSV]
+        ActionUpdate -> [CTApplicationJSON, CTTextCSV]
+        ActionDelete -> [CTApplicationJSON, CTTextCSV]
+        ActionInvoke -> [CTApplicationJSON]
+        ActionInspect -> [CTOpenAPI]
+        _ -> []
```

Comment:
Maybe you should handle `ActionInfo` and `ActionInappropriate` explicitly and remove the wildcard match. That way any new future actions will give a compiler warning of non-exhaustive pattern match.